packages/cloudflare/src/cli/build/patches/ast/util.ts

```diff
@@ -0,0 +1,48 @@
+import { type Edit, type NapiConfig, type SgNode } from "@ast-grep/napi";
+import yaml from "yaml";
+
+/**
+ * Returns the `Edit`s for an ast-grep rule in yaml format
+ *
+ * The rule must have a `fix` to rewrite the matched node.
+ *
+ * Tip: use https://ast-grep.github.io/playground.html to create rules.
+ *
+ * @param yamlRule The rule in yaml format
+ * @param root The root node
+ * @param once only apply once
+ * @returns A list of edits.
+ */
+export function applyRule(yamlRule: string, root: SgNode, { once = false } = {}) {
+  const rule: NapiConfig & { fix?: string } = yaml.parse(yamlRule);
+  if (rule.transform) {
+    throw new Error("transform is not supported");
+  }
+  if (!rule.fix) {
+    throw new Error("no fix to apply");
+  }
+
+  const fix = rule.fix;
+
+  const matches = once ? [root.find(rule)].filter((m) => m !== null) : root.findAll(rule);
+
+  const edits: Edit[] = [];
+
+  matches.forEach((match) => {
+    edits.push(
+      match.replace(
+        // Replace known placeholders by their value
+        fix
+          .replace(/\$\$\$([A-Z0-9_]+)/g, (_m, name) =>
+            match
+              .getMultipleMatches(name)
+              .map((n) => n.text())
+              .join()
```

Comment:
is it possible that you would get multiple matches for the placeholder name in the fix here? If so, is it right that all the matches are joined together when being replaced? I'm just wondering if you might end up with an edge case that results in incorrect values here due to multiple matches for the placeholder.

Reply:
ast-grep has placeholders for single node `$SINGLE` or multiple nodes `$$$MULTI` see [meta-variables](https://ast-grep.github.io/guide/pattern-syntax.html#meta-variable).

If your question is whether the variable can be re-used multiple times in the rule, then [yes](https://ast-grep.github.io/guide/pattern-syntax.html#meta-variable-capturing) but it would capture the same thing so it would like serialize to this one thing only.

Does it answer the question?

Reply:
Those links were helpful in understanding what's going on.

So what I was trying to say would probably be something like this example:
if your code contained `t = require('caniuse-lite', 'test');`, I believe the output would be incorrect - testing locally gives me `t = require('caniuse-lite',,,'test');`.

I know, that's not a valid require statement anyway because it's only supposed to have one arg. For this patch, it doesn't matter, but for other patches I think how the multi matches are replaced could potentially cause issues. I'm not quite sure, but happy to leave that for now

Reply:
Oh you are absolutely right, I should have used `.join("")`, I keep forgetting that `,` is the default :(

Reply:
it might be useful if we start writing unit tests for these patches to try and catch things like that

Reply:
I agree

<img width="726" alt="image" src="https://github.com/user-attachments/assets/90ecf9fe-6c42-45df-b330-dc3d9b64c85b" />

I will start by creating a P2 issue